File: e2e/python/sse_notifications_server.py
#!/usr/bin/env python3
"""SSE notifications server using Spikard SSE support"""

import asyncio

from spikard import Spikard
from spikard.config import ServerConfig
from spikard.sse import SseEvent, SseEventProducer


class NotificationProducer(SseEventProducer):
    """SSE event producer for notifications"""

    def __init__(self):
        self.count = 0
        self.max_events = 10

    async def next_event(self) -> SseEvent | None:
        """Generate the next notification event"""
        if self.count >= self.max_events:
            print("[NotificationProducer] Reached max events, ending stream")
            return None

        # Wait 1 second between events
        await asyncio.sleep(1)

        # Create notification event
        event = SseEvent(
            data={
                "notification": f"Notification {self.count}",
                "timestamp": self.count,
                "priority": "normal",
            },
            event_type="notification",
            id=str(self.count),
        )

        print(f"[NotificationProducer] Sending event {self.count}")
        self.count += 1
        return event

    async def on_connect(self) -> None:
        """Called when a client connects"""
        print("[NotificationProducer] Client connected to SSE stream")

    async def on_disconnect(self) -> None:
        """Called when a client disconnects"""
        print(f"[NotificationProducer] Client disconnected (sent {self.count} events)")


# Create Spikard app
app = Spikard()


# Register SSE endpoint
@app.sse("/notifications")
def notifications_endpoint():
    """SSE endpoint for notifications"""
    return NotificationProducer()


if __name__ == "__main__":
    print("Starting SSE notifications server on http://localhost:8000/notifications")
    config = ServerConfig(host="0.0.0.0", port=8000)
    app.run(config=config)

File: e2e/ruby/sse_notifications_server.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# SSE notifications server using Spikard SSE support

# Add packages/ruby to load path
$LOAD_PATH.unshift(File.expand_path('../../packages/ruby/lib', __dir__))

require 'spikard'

# SSE event producer for notifications
class NotificationProducer < Spikard::SseEventProducer
  def initialize
    super
    @count = 0
    @max_events = 10
  end

  def next_event
    if @count >= @max_events
      puts '[NotificationProducer] Reached max events, ending stream'
      return nil
    end

    # Wait 1 second between events
    sleep 1

    # Create notification event
    event = Spikard::SseEvent.new(
      data: {
        notification: "Notification #{@count}",
        timestamp: @count,
        priority: 'normal'
      },
      event_type: 'notification',
      id: @count.to_s
    )

    puts "[NotificationProducer] Sending event #{@count}"
    @count += 1
    event
  end

  def on_connect
    puts '[NotificationProducer] Client connected to SSE stream'
  end

  def on_disconnect
    puts "[NotificationProducer] Client disconnected (sent #{@count} events)"
  end
end

# Create Spikard app
app = Spikard::App.new

# Register SSE endpoint
app.sse('/notifications') do
  NotificationProducer.new
end

# Run server
if __FILE__ == $PROGRAM_NAME
  puts 'Starting SSE notifications server on http://localhost:8000/notifications'
  config = Spikard::ServerConfig.new(host: '0.0.0.0', port: 8000)
  app.run(config: config)
end
